File: pocketgull_flutter/lib/services/orcid_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:http/http.dart' as http;
import '../models/orcid_profile.dart';

class OrcidService with ChangeNotifier {
  String? _orcidId;
  OrcidProfile? _profile;
  bool _isLoading = false;
  String? _error;

  String? get orcidId => _orcidId;
  OrcidProfile? get profile => _profile;
  bool get isLoading => _isLoading;
  String? get error => _error;

  bool get isConnected => _orcidId != null;

  String _getBaseUrl() {
    if (kIsWeb) {
      final baseUri = Uri.base;
      return '${baseUri.scheme}://${baseUri.host}:${baseUri.port}';
    }
    // If Android emulator, use 10.0.2.2. If iOS/desktop, use localhost.
    if (defaultTargetPlatform == TargetPlatform.android) {
      return 'http://10.0.2.2:4000';
    }
    return 'http://localhost:4000';
  }

  Future<void> initialize() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final savedId = prefs.getString('orcid_id');
      if (savedId != null) {
        _orcidId = savedId;
        notifyListeners();
        await _fetchProfile(savedId);
      }
    } catch (e) {
      debugPrint('Failed to initialize OrcidService: $e');
    }
  }

  Future<bool> connectOrcid(String id) async {
    final cleanId = id.trim().replaceAll(RegExp(r'https?://orcid\.org/'), '');
    if (!RegExp(r'^\d{4}-\d{4}-\d{4}-\d{3}[0-9X]$').hasMatch(cleanId)) {
      _error = 'Invalid ORCID iD format. Expected: 0000-0002-1825-0097';
      notifyListeners();
      return false;
    }

    _isLoading = true;
    _error = null;
    notifyListeners();

    try {
      final baseUrl = _getBaseUrl();
      final response = await http.get(Uri.parse('$baseUrl/api/orcid/$cleanId'));

      if (response.statusCode == 200) {
        final rawJson = json.decode(response.body);
        final parsed = _parseOrcidData(cleanId, rawJson);

        _orcidId = cleanId;
        _profile = parsed;
        
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString('orcid_id', cleanId);
        
        _isLoading = false;
        notifyListeners();
        return true;
      } else {
        String serverError = 'Failed to fetch profile from ORCID.';
        try {
          final errorJson = json.decode(response.body);
          if (errorJson['error'] != null) {
            serverError = errorJson['error'];
          }
        } catch (_) {}
        _error = serverError;
      }
    } catch (e) {
      debugPrint('Failed to load ORCID profile: $e');
      _error = 'Failed to connect to ORCID server.';
    } finally {
      _isLoading = false;
      notifyListeners();
    }
    return false;
  }

  void disconnect() async {
    _orcidId = null;
    _profile = null;
    _error = null;
    notifyListeners();

    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.remove('orcid_id');
    } catch (e) {
      debugPrint('Failed to save ORCID disconnect state: $e');
    }
  }

  Future<void> _fetchProfile(String id) async {
    try {
      final baseUrl = _getBaseUrl();
      final response = await http.get(Uri.parse('$baseUrl/api/orcid/$id'));

      if (response.statusCode == 200) {
        final rawJson = json.decode(response.body);
        _profile = _parseOrcidData(id, rawJson);
        notifyListeners();
      } else {
        _error = 'Failed to load saved ORCID profile.';
        notifyListeners();
      }
    } catch (e) {
      debugPrint('Failed to auto-refresh ORCID profile: $e');
      _error = 'Failed to load saved ORCID profile.';
      notifyListeners();
    }
  }

  OrcidProfile _parseOrcidData(String orcidId, Map<String, dynamic> raw) {
    final person = raw['person'];
    final nameMap = person?['name'];
    final givenNames = nameMap?['given-names']?['value'] ?? '';
    final familyName = nameMap?['family-name']?['value'] ?? '';
    final nameList = [givenNames, familyName].where((s) => s.isNotEmpty).toList();
    final name = nameList.isEmpty ? 'Unknown Researcher' : nameList.join(' ');

    // Keywords
    final keywordsRaw = person?['keywords']?['keyword'] ?? [];
    final List<String> keywords = [];
    if (keywordsRaw is List) {
      for (final k in keywordsRaw) {
        final content = k['content'];
        if (content != null && content is String) {
          keywords.add(content);
        }
      }
    }

    // URLs
    final urlsRaw = person?['researcher-urls']?['researcher-url'] ?? [];
    final List<OrcidUrl> urls = [];
    if (urlsRaw is List) {
      for (final u in urlsRaw) {
        final uName = u['url-name'] ?? 'Website';
        final uVal = u['url']?['value'];
        if (uVal != null && uVal is String && uVal.isNotEmpty) {
          urls.add(OrcidUrl(name: uName, url: uVal));
        }
      }
    }

    // Works
    final worksRaw = raw['activities-summary']?['works']?['group'] ?? [];
    final List<OrcidWork> works = [];
    if (worksRaw is List) {
      for (final group in worksRaw) {
        final summaries = group['work-summary'];
        if (summaries is List && summaries.isNotEmpty) {
          final summary = summaries[0];
          final title = summary['title']?['title']?['value'] ?? 'Untitled Work';
          final url = summary['url']?['value'];
          final type = summary['type'];
          final year = summary['publication-date']?['year']?['value'];
          works.add(OrcidWork(
            title: title,
            url: url,
            type: type,
            year: year,
          ));
        }
      }
    }

    return OrcidProfile(
      orcidId: orcidId,
      name: name,
      keywords: keywords,
      works: works,
      urls: urls,
    );
  }
}

File: src/services/orcid.service.ts
import { Injectable, inject, signal, computed } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { SecureStorageService } from './secure-storage.service';

export interface IOrcidWork {
  title: string;
  url?: string;
  type?: string;
  year?: string;
}

export interface IOrcidProfile {
  orcidId: string;
  name: string;
  keywords: string[];
  works: IOrcidWork[];
  urls: { name: string; url: string }[];
}

@Injectable({
  providedIn: 'root'
})
export class OrcidService {
  private http = inject(HttpClient);
  private storage = inject(SecureStorageService);

  readonly orcidId = signal<string | null>(null);
  readonly orcidProfile = signal<IOrcidProfile | null>(null);
  readonly isLoading = signal<boolean>(false);
  readonly error = signal<string | null>(null);

  // Computed helper to check if connected
  readonly isConnected = computed(() => !!this.orcidId());

  constructor() {
    const savedId = this.storage.getItem('orcid_id');
    if (savedId) {
      this.orcidId.set(savedId);
      this.fetchProfile(savedId);
    }
  }

  /**
   * Connects an ORCID iD and fetches the public record
   */
  async connectOrcid(id: string): Promise<boolean> {
    const cleanId = id.trim().replace(/https?:\/\/orcid\.org\//, '');
    if (!/^\d{4}-\d{4}-\d{4}-\d{3}[0-9X]$/.test(cleanId)) {
      this.error.set('Invalid ORCID iD format. Expected: 0000-0002-1825-0097');
      return false;
    }

    this.isLoading.set(true);
    this.error.set(null);

    try {
      const data = await this.http.get<any>(`/api/orcid/${cleanId}`).toPromise();
      const parsed = this.parseOrcidData(cleanId, data);
      
      this.orcidId.set(cleanId);
      this.orcidProfile.set(parsed);
      
      this.storage.setItem('orcid_id', cleanId);
      return true;
    } catch (err: any) {
      console.error('Failed to load ORCID profile:', err);
      this.error.set(err.error?.error || 'Failed to fetch profile from ORCID.');
      return false;
    } finally {
      this.isLoading.set(false);
    }
  }

  /**
   * Disconnects ORCID connection
   */
  disconnect() {
    this.orcidId.set(null);
    this.orcidProfile.set(null);
    this.error.set(null);
    this.storage.removeItem('orcid_id');
  }

  private async fetchProfile(id: string) {
    try {
      const data = await this.http.get<any>(`/api/orcid/${id}`).toPromise();
      const parsed = this.parseOrcidData(id, data);
      this.orcidProfile.set(parsed);
    } catch (err: any) {
      console.error('Failed to auto-refresh ORCID profile:', err);
      this.error.set('Failed to load saved ORCID profile.');
    }
  }

  /**
   * Parse the ORCID JSON record response into a clean clinical context model
   */
  private parseOrcidData(orcidId: string, raw: any): IOrcidProfile {
    const person = raw?.person;
    const givenNames = person?.name?.['given-names']?.value || '';
    const familyName = person?.name?.['family-name']?.value || '';
    const name = `${givenNames} ${familyName}`.trim() || 'Anonymous Researcher';

    const keywords: string[] = [];
    if (person?.keywords?.keyword && Array.isArray(person.keywords.keyword)) {
      for (const k of person.keywords.keyword) {
        if (k?.content) keywords.push(k.content);
      }
    }

    const urls: { name: string; url: string }[] = [];
    if (person?.['researcher-urls']?.['researcher-url'] && Array.isArray(person['researcher-urls']['researcher-url'])) {
      for (const u of person['researcher-urls']['researcher-url']) {
        if (u?.url?.value) {
          urls.push({ name: u['url-name'] || u.url.value, url: u.url.value });
        }
      }
    }

    const works: IOrcidWork[] = [];
    const groups = raw?.['activities-summary']?.works?.group;
    if (groups && Array.isArray(groups)) {
      for (const group of groups) {
        const summaries = group?.['work-summary'];
        if (summaries && Array.isArray(summaries) && summaries.length > 0) {
          const work = summaries[0]; // Get primary summary
          const title = work?.title?.title?.value;
          if (title) {
            works.push({
              title,
              url: work.url?.value,
              type: work.type,
              year: work['publication-date']?.year?.value
            });
          }
        }
      }
    }

    return { orcidId, name, keywords, works, urls };
  }
}
